src/pudl/analysis/service_territory.py

```diff
@@ -0,0 +1,605 @@
+                "county_name_census",
+                "state",
+                "state_id_fips"
+            ], axis="columns")
+            .reset_index()
+        )
+    return out_gdf
+
+
+def utility_geometries(ids,
+                       st_eia861,
+                       ba_assn_eia861,
+                       census_gdf,
+                       dissolve=False,
+                       limit_by_state=False):
+    """
+    Compile utility territory geometries based on county_id_fips.
+
+    Compile geometries for all of the utilities indicated by the ``utility_id_eia``
+    values specified by ``ids``. Use the EIA 861 Service Territory table to look up the
+    list of county FIPS IDs aassociated with each utility in each year, and then merge
+    in the corresponding county geometries from the US Census DP1 data.
+
+    Optionally dissolve all of the county level geometries into a single geometry for
+    each utility-year.
+
+    Note:
+        Dissolving geometires is a costly operation, and may take half an hour or more
+        if you are processing all utilities for all years. Dissolving also means that
+        all the per-county information will be lost, rendering the output inappropriate
+        for use in many analyses. Dissolving is mostly useful for generating
+        visualizations.
+
+    Args:
+        ids (iterable of ints): The set of utility_id_eia values for which we are
+            compiling geometries.
+        st_eia861 (pandas.DataFrame): The service_territory_eia861 table.
+        ba_assn_eia861 (pandas.DataFrame): The balancing_authority_assn_eia861 table.
+        census_gdf (geopandas.GeoDataFrame): The US Census DP1 county-level geometries
+            as returned by get_census2010_gdf().
+        dissolve (bool): If False, each record in the compiled territory will correspond
+            to a single county, with a county-level geometry, and there will be many
+            records enumerating all the counties associated with a given utility_id_eia
+            in each year. If dissolve=True, all of the county-level geometries for each
+            utility in each year will be merged together ("dissolved") resulting in a
+            single geometry and record for each utility-year.
+        limit_by_state (bool): If True, limit the included counties to only those for
+            which data has actually been reported for that utility in association with
+            the state that the county is part of.
+
+    Returns:
+        geopandas.GeoDataFrame
+
+    """
+    if limit_by_state:
+        raise NotImplementedError(
+            "Limit-by-State not yet implemented for utility service territories."
+        )
+
+    util_gdf = (
+        utility_counties(
+            util_ids=ids,
+            st_eia861=st_eia861
+        )
+        .pipe(
+            add_geometries,
+            census_gdf,
+            dissolve=dissolve,
+            dissolve_by=["report_date", "utility_id_eia"]
+        )
+    )
```

Comment:
I almost made a single function for both of them, but yeah, they currently take different arguments, and they have different options,

Reply:
I'm not sure I'm understanding what it is you're getting at here exactly. Can you give me some more details?

Reply:
the `utility_geometries` and `balancing_authority_geometries` functions look like they both are doing the same analogous two things (call `whatever_counties` and call `compile_geoms`). What I was wondering is whether or not these two functions are really necessary - if this can be looped into `compile_geoms`. Not a big deal either way.

Reply:
Ah, I see. I wanted to have these functions available independently for interactive use where you don't actually want to compile *all* of the possible geometries for every single balancing authority or utility -- `compile_geoms` generates all the geometries for all entities in all years, whereas the individual `_geometries` functions take a list of specific IDs, and a dataframe with a potentially limited set of years of data (for service territories, etc.).